File: backend-go/internal/domain/engine/component_bindings.go
package engine

import "github.com/tiagofur/muebles-backend/internal/domain"

// ApplyEvaluatedComponentBindings copies component quantity/condition consumers.
// Callers must validate definitions, consumer targets and evaluated scalar values
// first. This is not a release readiness gate or relationship materializer.
func ApplyEvaluatedComponentBindings(module domain.Module, values map[string]any) domain.Module {
	for _, definition := range module.ParameterDefinitions {
		binding := definition.Binding
		if binding == nil || (binding.Kind != domain.FurnitureParameterBindingComponentQuantity && binding.Kind != domain.FurnitureParameterBindingComponentCondition) {
			continue
		}
		quantity := -1
		if binding.Kind == domain.FurnitureParameterBindingComponentQuantity {
			value, ok := values[definition.Name].(float64)
			if !ok {
				continue
			}
			quantity = int(value)
		} else {
			value, ok := values[definition.Name].(bool)
			if !ok {
				continue
			}
			if !value {
				quantity = 0
			}
		}
		updated := make([]domain.ComponentInstance, 0, len(module.Components))
		for _, instance := range module.Components {
			if instance.ComponentID == binding.ComponentID {
				if quantity == 0 {
					continue
				}
				if quantity > 0 {
					instance.Quantity = quantity
				}
			}
			updated = append(updated, instance)
		}
		module.Components = updated
	}
	return module
}

File: packages/domain/src/engine/componentBindings.ts
import type { FurnitureParameter } from '../smartFurnitureDomain';
import type { FurnitureParameterValue } from '../furnitureParameters';
import type { Module } from '../types';

/** Mirrors Go ApplyEvaluatedComponentBindings, not a release readiness gate.
 * Callers must validate definitions, consumer targets and evaluated values first.
 * Relationship effects are outside this component-only materialization helper.
 */
export function applyEvaluatedComponentBindings(
  module: Module,
  definitions: readonly FurnitureParameter[],
  values: Readonly<Record<string, FurnitureParameterValue>>,
): Module {
  let prepared = module;
  for (const definition of definitions) {
    const binding = definition.binding;
    if (!binding || !['componentQuantity', 'componentCondition'].includes(binding.kind)) continue;
    const value = values[definition.name];
    let quantity = -1;
    if (binding.kind === 'componentQuantity') {
      if (typeof value !== 'number') continue;
      quantity = Math.trunc(value);
    } else {
      if (typeof value !== 'boolean') continue;
      if (!value) quantity = 0;
    }
    prepared = {
      ...prepared,
      components: (prepared.components ?? []).flatMap((instance) => {
        if (instance.componentId !== binding.componentId) return [instance];
        if (quantity === 0) return [];
        return [quantity > 0 ? { ...instance, quantity } : instance];
      }),
    };
  }
  return prepared;
}
